go/pkg/common/utils.go

```diff
@@ -124,3 +125,32 @@ func SyncMapLength(m *sync.Map) int {
 	})
 	return i
 }
+
+// RetryUntilSuccessful calls callback every interval until duration until it returns false (to not retry)
+func RetryUntilSuccessful(duration time.Duration,
+	interval time.Duration,
+	callback func() (bool, error)) error {
+	var lastErr error
+	timedOutErrorMessage := "Timed out waiting until successful"
+	deadline := time.Now().Add(duration)
+
+	// while we haven't passed the deadline
+	for !time.Now().After(deadline) {
+		shouldRetry, err := callback()
+		lastErr = err
+		if !shouldRetry {
+			return err
+		}
+		time.Sleep(interval)
+		continue
+
+	}
+	if lastErr != nil {
+
+		// wrap last error
+		return errors.Wrapf(lastErr, timedOutErrorMessage)
+	}
```

Comment:
if try 2 failed and try 3 succeeded, won't this cause us to receive an error even though it succeeded?

Reply:
If the last try succeeds, then `lastErr == nil`.
`lastErr` always gets overridden by the error returned from the callback.